tests/sources/core/test_memory_utils.cpp

```diff
@@ -144,3 +147,119 @@ TEST_CASE("memcpy_bits", "[core][memory]")
+{
+	CHECK_THROWS(safe_static_cast<uint32_t>(SignedEnum::I32_MIN));
+	CHECK_THROWS(safe_static_cast<uint32_t>(SignedEnum::I16_MIN));
+	CHECK(safe_static_cast<uint32_t>(SignedEnum::I16_MAX) == INT16_MAX);
+	CHECK(safe_static_cast<uint32_t>(SignedEnum::I32_MAX) == INT32_MAX);
+
+	CHECK(safe_static_cast<int32_t>(SignedEnum::I32_MIN) == INT32_MIN);
+	CHECK(safe_static_cast<int32_t>(SignedEnum::I16_MIN) == INT16_MIN);
+	CHECK(safe_static_cast<int32_t>(SignedEnum::I16_MAX) == INT16_MAX);
+	CHECK(safe_static_cast<int32_t>(SignedEnum::I32_MAX) == INT32_MAX);
+
+	CHECK_THROWS(safe_static_cast<uint16_t>(SignedEnum::I32_MIN));
+	CHECK_THROWS(safe_static_cast<uint16_t>(SignedEnum::I16_MIN));
+	CHECK(safe_static_cast<uint16_t>(SignedEnum::I16_MAX) == INT16_MAX);
+	CHECK_THROWS(safe_static_cast<uint16_t>(SignedEnum::I32_MAX));
+
+	CHECK_THROWS(safe_static_cast<int16_t>(SignedEnum::I32_MIN));
+	CHECK(safe_static_cast<int16_t>(SignedEnum::I16_MIN) == INT16_MIN);
+	CHECK(safe_static_cast<int16_t>(SignedEnum::I16_MAX) == INT16_MAX);
+	CHECK_THROWS(safe_static_cast<int16_t>(SignedEnum::I32_MAX));
+}
+
+TEST_CASE("safe_static_cast from signed int", "[core][memory]")
+{
+	CHECK_THROWS(safe_static_cast<uint32_t>(INT32_MIN));
+	CHECK_THROWS(safe_static_cast<uint32_t>(INT16_MIN));
+	CHECK(safe_static_cast<uint32_t>(INT16_MAX) == INT16_MAX);
+	CHECK(safe_static_cast<uint32_t>(INT32_MAX) == INT32_MAX);
+
+	CHECK(safe_static_cast<int32_t>(INT32_MIN) == INT32_MIN);
+	CHECK(safe_static_cast<int32_t>(INT16_MIN) == INT16_MIN);
+	CHECK(safe_static_cast<int32_t>(INT16_MAX) == INT16_MAX);
+	CHECK(safe_static_cast<int32_t>(INT32_MAX) == INT32_MAX);
+
+	CHECK_THROWS(safe_static_cast<uint16_t>(INT32_MIN));
+	CHECK_THROWS(safe_static_cast<uint16_t>(INT16_MIN));
+	CHECK(safe_static_cast<uint16_t>(INT16_MAX) == INT16_MAX);
+	CHECK_THROWS(safe_static_cast<uint16_t>(INT32_MAX));
+
+	CHECK_THROWS(safe_static_cast<int16_t>(INT32_MIN));
+	CHECK(safe_static_cast<int16_t>(INT16_MIN) == INT16_MIN);
+	CHECK(safe_static_cast<int16_t>(INT16_MAX) == INT16_MAX);
+	CHECK_THROWS(safe_static_cast<int16_t>(INT32_MAX));
+}
+
+TEST_CASE("safe_static_cast from unsigned int", "[core][memory]")
+{
+	CHECK(safe_static_cast<uint32_t>(0U) == 0);
+	CHECK(safe_static_cast<uint32_t>(UINT16_MAX) == UINT16_MAX);
+	CHECK(safe_static_cast<uint32_t>(UINT32_MAX) == UINT32_MAX);
+
+	CHECK(safe_static_cast<int32_t>(0U) == 0);
+	CHECK(safe_static_cast<int32_t>(UINT16_MAX) == UINT16_MAX);
+	CHECK_THROWS(safe_static_cast<int32_t>(UINT32_MAX));
+
+	CHECK(safe_static_cast<uint16_t>(0U) == 0);
+	CHECK(safe_static_cast<uint16_t>(UINT16_MAX) == UINT16_MAX);
+	CHECK_THROWS(safe_static_cast<uint16_t>(UINT32_MAX));
+
+	CHECK(safe_static_cast<int16_t>(0U) == 0);
+	CHECK_THROWS(safe_static_cast<int16_t>(UINT16_MAX));
+	CHECK_THROWS(safe_static_cast<int16_t>(UINT32_MAX));
+}
+
+#ifdef _MSC_VER
+	#pragma warning(push)
+	#pragma warning(disable: 4756) /* overflow in constant arithmetic */
```

Comment:
what generates this warning?

Reply:
VS 2015 and the "safe_static_cast from double" test. It's a compile time test of static_cast<float>(double) when the input is known to be a constant. Doesn't do this for integers though!